kntest/pkg/cluster/README.md

```diff
@@ -0,0 +1,58 @@
+## prow-cluster-operation
+
+`kntest cluster` command is used for creating, deleting or getting a GKE cluster
```

Comment:
let's consider adding another subcommand
```
kntest cluster gke
```

Reply:
It needs to be a single command that chooses based on environment variables.

Otherwise, we have to write a bunch of bash to decide which cluster command to call.... that's the whole thing we're trying to avoid.

Reply:
Even the cluster provider can be an environment variable, e.g. the command can be called with `kntest cluster ${PROVIDER}`, so it doesn't matter how we provide the commands IMO.

Reply:
I'm going to support both env vars and flags, the kntest command will use the env vars by default, but users can also use flags to overwrite them. Does it sound a reason direction to you guys?

Reply:
I'm OK kntest supporting some env vars, but hesitating on extensively using env vars. Do we have a boundary here? Also brings in more opinions @mattmoor

Reply:
I'm going to withdraw what I said above - in this PR I'm not going to add any support for env vars in `kntest`, I'm going to support `kntest cluster gke` as a subcommand though.

My point is the current `kntest` design is focusing on simplifying the CI/CD flow we currently have, and it's still not clear how we want to support the downstream processes. So I'd rather not adding it at this point but leave it to the next step of `kntest`.

Reply:
I don't support this at all. If we configure things via arguments, then we have to use bash scripts to configure and pass things in as arguments -- the only things we can set in the Prow Job config is environment variables or args to the script, which then has to parse.

There must not be a subcommand or we aren't fixing the problem -- the complexity is left in the bash script. If we're only supporting one then we don't need a subcommand.

We either support environment variables or a config file. Nothing else takes the complexity of decision making out of bash.

Reply:
Honestly I don't see why subcommand is a problem in the context you provided. Like I said above we can run the command like `kntest cluster ${PROVIDER}`, and `PROVIDER` can be an env var in both approaches regardless whether we support subcommands or not.

However, the drawback of not having subcommands is clear: they will share all the flags (like 
the command to create `kind` cluster will also have the `gcp-project` flag), we cannot define and set clear ownership on these commands. These are also the issues [kubetest2](https://github.com/kubernetes/test-infra/tree/master/kubetest2) is trying to avoid.

Regarding the env vars, I definitely agree it's an option, but just as you said, we are still not sure whether we should support environment variables or a config file. So at this stage I'd like to first get rid of the huge bash functions in https://github.com/knative/test-infra/blob/master/scripts/e2e-tests.sh#L185-L321, and we can always iterate on it when we get a clearer picture.

Reply:
I'm okay with having flags if we will soon support the exact same options as env var/config file.

Otherwise, we're going to have this in our bash script:

```
if [[${PROVIDER}="gke"]]; then
kntest cluster gke --max-nodes ${GKE_MAX_NODES} --min-nodes ${GKE_MIN_NODES} --project-name=${GKE_PROJECT_ID} etc
elif [[${PROVIDER}="kind"]]; then
kntest cluster kind --other-kind-option ${OTHER_KIND_ENV} --another-kind-option ${ANOTHER_KIND_ENV}
elif [[${PROVIDER=...etc...
fi
```

Let's just solve the problem completely. `kntest cluster create`. That should be the only line and the complete command in the script.

We have/had this same problem with namespace being hardcoded in e2e_flags.go. Don't make me hack shit to get it done -- make the only available option the right one.

Reply:
I agree that `kntest cluster create` makes some sense, but does `kntest cluster ${PROVIDER} create` indeed add too much complexity?

Reply:
The switch statement Cory mentioned above doesn't seem to be avoidable in advance usage of `kntest` no matter which direction we go, as provider specific logics differ so much from each other, for example `--project` flag is a gcloud only logic, it's meaningless in all other platforms. But in default cases that shouldn't be an issue, caller should expect calling `kntest cluster ${PROVIDER}` and get a default cluster